doc/man/gnunet-vectorproduct.1
.TH GNUNET\-VECTORPRODUCT 1 "8 Aug 2013" "GNUnet"

.SH NAME
gnunet\-vectorproduct \- compute a vectorproduct

.SH SYNOPSIS
.B gnunet\-vectorproduct
.RI [ options ]
.br

.SH DESCRIPTION
\fBgnunet-vectorproduct\fP enables you to compute a vectorproduct across two peers \fBAlice\fP and \fBBob\fP.

A client can issue one of two messages to its service:
.TS
tab (@);
l lx.
1@T{
A request to compute a vectorproduct with another peer (\fBAlice\fP) 
T}
2@T{
Elements to support a peer in computing a vectorproduct (\fBBob\fP) 
T}
.TE

Both requests must share the same SID, which can be an arbitrary string identifying the session. SIDs should be unique, however it is sufficient to guarantee the uniqueness of the tupel element count and session ID.

\fBAlice\fP\'s client must supply the ASCII encoded peer ID of bob\'s service, it will internally be checked by the client for validity. Invalid values here result in the client or the service failing the session.

Elements are handed over as signed decimal integers, the element count supplied by \fBAlice\fP and \fBBob\fP must match. \fBAlice\fP can also supply a mask for these values to her service, which allows partial vector products to be computed across the vector. Elements can be masked by setting their the corresponding mask element to zero, any other value means the element will not be masked. \fBAlice\fP\'s client will also mask all 0-values to avoid information leakage to \fBBob\fP. 

The protocol by definition relies on \fBAlice\fP and \fBBob\fP being benign, thus \fBBob\fP can arbitrarily falsify his information. Both peers collaborate to achieve a correct result. 

.SH OPTIONS
.B
.IP "\-e ELEMENTS, \-\-elements=ELEMENTS"
The element-vector the vectorproduct should be computed over in signed decimal form, eg: \"42,1,-3,3,7\". Zero value elements will be automatically masked.
.B
.IP "\-m MASK, \-\-mask=MASK"
Elements in the vector can be masked. There must be at least two elements left in the vector to compute a vectorproduct. Non-Zero values indicate an element is not maskes.
.B
.IP "\-k KEY, \-\-key=KEY"
The session key, a shared string of arbitrary length from which the SID will be generated
.B
.IP "\-c FILENAME,  \-\-config=FILENAME"
Use the configuration file FILENAME.
.B
.IP "\-p PEERID, \-\-peer=PEERID"
The remote peer\'s ASCII-armored gnunet-peer ID as output by gnunet-peerinfo. If this option is not given, the peer will take the \fBBob\fP\'s role.
.B
.IP "\-h, \-\-help"
Print short help on options.
.B
.IP "\-L LOGLEVEL, \-\-loglevel=LOGLEVEL"
Use LOGLEVEL for logging.  Valid values are DEBUG, INFO, WARNING and ERROR.
.B
.IP "\-v, \-\-version"
Print GNUnet version number.


.SH BUGS
Report bugs by using Mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>

.SH SEE ALSO
gnunet\-peerinfo(1)


ChangeLog
Mon Oct 31 18:46:36 CET 2011
	Releasing GNUnet 0.9.0pre4.  This release includes a working
	implementation of mesh (unoptimized), a working implementation
	of the VPN (could use more testing), a working implementation 
	of the WLAN transport (still very slow, unoptimized) and
	many, many bugfixes all over the place.  The major known
	issues are:

	* TESTING is known to sometimes crash on shutdown
	* DV is known to crash / not work properly
	* CHAT is experimental and was not reviewed yet

	More details to individual bugs that were fixed in this
	release are documented at:
	https://gnunet.org/bugs/changelog_page.php?version_id=182

Mon Sep 19 14:16:22 CEST 2011
	Releasing GNUnet 0.9.0pre3.  Instead of listing all that has
	changed, here is the short list of what is known NOT to work,
	in order of minor flaws to major issues, by module:

	* FS still has a few missing features, but largely works
	* TRANSPORT ATS is not used for actual bandwidth assignment,
	          important options to make it truly useful are still missing
	* TESTING does not support incremental HELLO changes, 
	          code has known shutdown issues (can crash)
	* VPN is experimental, but has been shown to work to some degree
	* MESH does not work (mock API emulates for VPN)
        * TRANSPORT-WLAN is experimental
        * TRANSPORT-HTTP is experimental, has known bugs
	* TRANSPORT-DV is experimental, has known bugs, protocol may still change
	* CHAT is experimental, has known bugs and was not reviewed yet

Thu Dec 23 23:33:05 CET 2010
	Releasing GNUnet 0.9.0pre2.

Thu Oct	 7 20:51:05 CEST 2010
	Removed dependency on GNU MP (GMP). The code now uses the libgcrypt MPI
	interface.

Sat Jul  3 20:47:45 CEST 2010
	Releasing GNUnet 0.9.0pre1.

Wed Jun 23 16:34:38 CEST 2010
	Added support for systemd-compatible passing of listen-sockets
	by ARM to services as well as systemd compatibility for gnunet-service-arm
	itself.  At least for non-MINGW systems this should work.

Sat Jun  5 18:08:39 CEST 2010
	Added support for UNIX domain sockets, code also defaults to
	them when available.

Sun May  2 13:49:10 CEST 2010
	Fixed problem with platform-dependence of format for IP addresses
	in HELLOs for TCP and UDP transport.
	Reduced address length field in HELLOs to 16 bit (was 32 bit).
	These changes break transport compatibility.

Fri Apr 16 18:19:05 CEST 2010
	Nearly complete rewrite and new overall architecture.  Many
	features are still missing, but basic system seems to be
	working again.  Maybe time to again track major changes
	in the ChangeLog.  Releasing 0.9.0pre0.


Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent.
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).

Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in
	util/network/dns.c; improved IPv6 support for
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.


src/vectorproduct/gnunet-vectorproduct.c
/*
     This file is part of GNUnet.
     (C) 2013 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 3, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
 */

/**
 * @file vectorproduct/gnunet-vectorproduct.c
 * @brief vectorproduct client
 * @author Christian M. Fuchs
 */
#define GCRYPT_NO_DEPRECATED
#include <gcrypt.h>
#include <inttypes.h>

#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_vectorproduct_service.h"
#include "gnunet_protocols.h"

#define LOG(kind,...) GNUNET_log_from (kind, "gnunet-vectorproduct",__VA_ARGS__)
/**
 * Option -p: destination peer identity for checking message-ids with
 */
static char *input_peer_id = NULL;

/**
 * Option -p: destination peer identity for checking message-ids with
 */
static char *input_key = NULL;

/**
 * Option -e: vector to calculate a vectorproduct with
 */
static char *input_elements = NULL;

/**
 * Option -m: message-ids to calculate a vectorproduct with
 */
static char *input_mask = NULL;

/**
 * the count of the messages sent to the service for processing
 */
static unsigned short element_count;

/**
 * the count of the mask bytes
 */
unsigned short mask_length = 0;

/**
 * the count of the number of mask bytes
 */
unsigned short mask_bytes;

/**
 * the array of converted message IDs to send to our service
 */
static int32_t * elements = NULL;

/**
 * the array of converted message IDs to send to our service
 */
static unsigned char * mask = NULL;

/**
 * information about the peer we are comparing with
 */
struct GNUNET_PeerIdentity peer;

/**
 * information about the peer we are comparing with
 */
struct GNUNET_HashCode key;

/**
 * Pointer to the GNUNET_VECTORPRODUCT_Handle
 */
struct GNUNET_VECTORPRODUCT_Handle *handle;

/**
 * Global return value
 */
static int ret;

struct GNUNET_VECTORPRODUCT_TestCls
{
  struct GNUNET_VECTORPRODUCT_Handle * h;
};

struct GNUNET_VECTORPRODUCT_TestCls test_cls;


/**
 * Callback called if we are initiating a new computation session
 * 
 * @param cls unused
 * @param status if our job was successfully processed 
 */
static void
responder_callback (void *cls,
                    const struct GNUNET_HashCode * key,
                    enum GNUNET_VECTORPRODUCT_ResponseStatus status)
{
  ret = -1;

  switch (status)
    {
    case GNUNET_VECTORPRODUCT_Status_Success:
      ret = 0;
      LOG (GNUNET_ERROR_TYPE_INFO, "Session %s concluded.\n", GNUNET_h2s (key));
      break;
    case GNUNET_VECTORPRODUCT_Status_InvalidResponse:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: invalid response\n", GNUNET_h2s (key));
      break;
    case GNUNET_VECTORPRODUCT_Status_Timeout:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: timeout\n", GNUNET_h2s (key));
      break;
    case GNUNET_VECTORPRODUCT_Status_Failure:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: service failure\n", GNUNET_h2s (key));
    case GNUNET_VECTORPRODUCT_Status_ServiceDisconnected:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: service disconnect!!\n", GNUNET_h2s (key));
      break;
    default:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: return code %d\n", GNUNET_h2s (key), (int) status);
    }

  GNUNET_VECTORPRODUCT_disconnect (handle);
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Callback called if we are initiating a new computation session
 * 
 * @param cls unused
 * @param key unused
 * @param peer unused
 * @param status if our job was successfully processed 
 * @param size size of the msg returned
 * @param msg the response we got.
 * @param type of the message received 
 */
static void
requester_callback (void *cls,
        const struct GNUNET_HashCode * key,
        const struct GNUNET_PeerIdentity * peer,
        enum GNUNET_VECTORPRODUCT_ResponseStatus status,
        const struct GNUNET_VECTORPRODUCT_client_response *msg)
{
  uint32_t product_len;
  ret = -1;

  switch (status)
    {
    case GNUNET_VECTORPRODUCT_Status_Success:
      product_len = ntohl (msg->product_length);

      LOG (GNUNET_ERROR_TYPE_INFO, "Session %s concluded.\n", GNUNET_h2s (key));

      if (0 < product_len && NULL != &msg[1])
        {
          gcry_mpi_t result;
          size_t read = 0;

          if (0 != gcry_mpi_scan (&result, GCRYMPI_FMT_USG, &msg[1], product_len, &read))
            LOG (GNUNET_ERROR_TYPE_ERROR, "Could not convert to mpi to value!\n");
          else
            {
              unsigned char * buf;
              gcry_mpi_aprint (GCRYMPI_FMT_HEX, &buf, NULL, result);

              printf ("Successfully computed result for session %s: %s\n", GNUNET_h2s (key), buf);
              ret = 0;
            }
        }
      else
        { //currently not used, but if we get more info due to MESH we will need this
          LOG (GNUNET_ERROR_TYPE_ERROR, "Service-side error in session %s, return code: %d\n", GNUNET_h2s (key), product_len);
        }
      break;
    case GNUNET_VECTORPRODUCT_Status_InvalidResponse:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: invalid response\n", GNUNET_h2s (key));
      break;
    case GNUNET_VECTORPRODUCT_Status_Timeout:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: timeout\n", GNUNET_h2s (key));
      break;
    case GNUNET_VECTORPRODUCT_Status_Failure:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: service failure\n", GNUNET_h2s (key));
    case GNUNET_VECTORPRODUCT_Status_ServiceDisconnected:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Disconnected from service.\n", GNUNET_h2s (key));
      break;
    default:
      LOG (GNUNET_ERROR_TYPE_ERROR, "Session %s failed: return code %d\n", GNUNET_h2s (key), (int) status);
    }
  GNUNET_VECTORPRODUCT_disconnect (handle);
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  char * begin = input_elements;
  char * end;
  int32_t element;
  int i;
  ret = -1;

  if (NULL == input_elements)
    {
      FPRINTF (stderr, "%s", _ ("You must specify at least one message ID to check!\n"));
      return;
    }

  if (NULL == input_key)
    {
      FPRINTF (stderr, "%s", _ ("This program needs a session identifier for comparing vectors.\n"));
      return;
    }

  if (1 > strnlen (input_key, sizeof (struct GNUNET_HashCode)))
    {
      FPRINTF (stderr, _ ("Please give a session key for --input_key!\n"));
      return;
    }
  GNUNET_CRYPTO_hash (input_key, strlen (input_key), &key);

  if (input_peer_id && GNUNET_OK != GNUNET_CRYPTO_hash_from_string (input_peer_id,
                                                                    (struct GNUNET_HashCode *) &peer))
    {
      FPRINTF (stderr, _ ("Tried to set initiator mode, as peer ID was given. "
                          "However, `%s' is not a valid peer identifier.\n"),
               input_peer_id);
      return;
    }

  int exit_loop = 0;
  /* Read input_elements_peer1, and put in elements_peer1 array */
  do
    {
      unsigned int mcount = element_count;
      //ignore empty rows of ,,,,,,
      while (*begin == ',')
        begin++;
      // get the length of the current element and replace , with null
      for (end = begin; *end && *end != ','; end++);

      if (*end == '\0')
        exit_loop = 1;

      if (*end == ',')
        *end = '\0';

      if (1 != sscanf (begin, "%" SCNd32, &element))
        {
          FPRINTF (stderr, _ ("Could not convert `%s' to int32_t.\n"), begin);
          return;
        }

      GNUNET_array_append (elements, mcount, element);
      element_count++;

      begin = ++end;
    }
  while (!exit_loop);

  GNUNET_assert (elements != NULL);
  GNUNET_assert (element_count > 1);
  mask_length = element_count / 8 + (element_count % 8 ? 1 : 0);
  mask = GNUNET_malloc ((element_count / 8) + 2);

  /* Read input_mask_peer1 and read in mask_peer1 array */
  if (NULL != input_mask)
    {
      begin = input_mask;
      unsigned short mask_count = 0;
      int exit_loop = 0;

      do
        {
          //ignore empty rows of ,,,,,,
          while (* begin == ',')
            begin++;
          // get the length of the current element and replace , with null
          // gnunet_ascii-armor uses base32, thus we can use , as separator!
          for (end = begin; *end && *end != ','; end++);

          if (*end == '\0')
            exit_loop = 1;

          if (*end == ',')
            *end = '\0';

          if (1 != sscanf (begin, "%" SCNd32, &element))
            {
              FPRINTF (stderr, _ ("Could not convert `%s' to int32_t.\n"), begin);
              return;
            }

          GNUNET_assert (mask_count <= element_count);

          if (element)
            mask[mask_count / 8] = mask[mask_count / 8] | 1 << (mask_count % 8);

          mask_count++;
          begin = ++end;
        }
      while (!exit_loop);
      // +1 to see if we would have more data, which would indicate malformed/superficial input
      GNUNET_assert (mask_count == element_count);
    }
  else if (input_peer_id)
    {
      for (i = 0; i <= mask_length; i++)
        mask[i] = UCHAR_MAX; // all 1's
    }

  handle = GNUNET_VECTORPRODUCT_connect (cfg);
  if (handle == NULL)
    {
      FPRINTF (stderr, _ ("Could not connect to the GNUNET Vector Product Service\n"));
      return;
    }

  test_cls.h = handle;

  if (input_peer_id && !GNUNET_VECTORPRODUCT_request (handle,
                                                      &key,
                                                      &peer,
                                                      element_count,
                                                      mask_length,
                                                      elements, mask,
                                                      GNUNET_TIME_UNIT_MINUTES,
                                                      &requester_callback,
                                                      (void *) &test_cls))
    return;
  if ( !input_peer_id && !GNUNET_VECTORPRODUCT_prepare_response (handle,
                                                   &key,
                                                   element_count,
                                                   elements,
                                                   GNUNET_TIME_UNIT_MINUTES,
                                                   &responder_callback,
                                                   (void *) &test_cls))
    return;

  ret = 0;
}


/**
 * The main function to the vectorproduct client.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  static const struct GNUNET_GETOPT_CommandLineOption options[] = {
    {'e', "elements", "\"val1,val2,...,valn\"",
      gettext_noop ("A comma separated list of elements to compare as vector with our remote peer."),
      1, &GNUNET_GETOPT_set_string, &input_elements},
    {'m', "mask", "\"0,1,...,maskn\"",
      gettext_noop ("A comma separated mask to select which elements should actually be compared."),
      1, &GNUNET_GETOPT_set_string, &input_mask},
    {'p', "peer", "PEERID",
      gettext_noop ("[Optional] peer to calculate our vectorproduct with. If this parameter is not given, the service will wait for a remote peer to compute the request."),
      1, &GNUNET_GETOPT_set_string, &input_peer_id},
    {'k', "key", "TRANSACTION_ID",
      gettext_noop ("Transaction ID shared with peer."),
      1, &GNUNET_GETOPT_set_string, &input_key},
    GNUNET_GETOPT_OPTION_END
  };

  return (GNUNET_OK ==
          GNUNET_PROGRAM_run (argc,
                              argv,
                              "gnunet-vectorproduct",
                              gettext_noop ("Calculate the Vectorproduct with a GNUnet peer."),
                              options, &run, NULL)) ? ret : 1;
}

